riscv/processor.cc

```diff
@@ -861,11 +1052,118 @@ void processor_t::set_csr(int which, reg_t val)
       mask |= 1L << ('F' - 'A');
       mask |= 1L << ('D' - 'A');
       mask |= 1L << ('C' - 'A');
+      mask |= 1L << ('H' - 'A');
       mask &= max_isa;
 
       state.misa = (val & mask) | (state.misa & ~mask);
       break;
     }
+    case CSR_HSTATUS: {
+      reg_t mask = HSTATUS_VTSR | HSTATUS_VTW | HSTATUS_VTVM |
+                   HSTATUS_HU | HSTATUS_SPVP | HSTATUS_SPV | HSTATUS_GVA;
+      state.hstatus = (state.hstatus & ~mask) | (val & mask);
+      break;
+    }
+    case CSR_HEDELEG: {
+      reg_t mask =
+        (1 << CAUSE_MISALIGNED_FETCH) |
+        (1 << CAUSE_BREAKPOINT) |
+        (1 << CAUSE_MISALIGNED_LOAD) |
+        (1 << CAUSE_LOAD_ACCESS) |
+        (1 << CAUSE_MISALIGNED_STORE) |
+        (1 << CAUSE_STORE_ACCESS) |
+        (1 << CAUSE_USER_ECALL) |
+        (1 << CAUSE_FETCH_PAGE_FAULT) |
+        (1 << CAUSE_LOAD_PAGE_FAULT) |
+        (1 << CAUSE_STORE_PAGE_FAULT);
+      state.hedeleg = (state.hedeleg & ~mask) | (val & mask);
+      break;
+    }
+    case CSR_HIDELEG: {
+      reg_t mask = MIP_VS_MASK;
+      state.hideleg = (state.hideleg & ~mask) | (val & mask);
+      break;
+    }
+    case CSR_HIE: {
+      reg_t mask = MIP_HS_MASK;
+      state.mie = (state.mie & ~mask) | (val & mask);
+      break;
+    }
+    case CSR_HTIMEDELTA:
+    case CSR_HTIMEDELTAH:
+      throw trap_illegal_instruction(0);
+      break;
+    case CSR_HCOUNTEREN:
+      state.hcounteren = val;
+      break;
+    case CSR_HGEIE:
+      /* Ignore */
+      break;
+    case CSR_HTVAL:
+      state.htinst = val;
+      break;
+    case CSR_HIP: {
+      reg_t mask = MIP_VSSIP;
+      state.mip = (state.mip & ~mask) | (val & mask);
+      break;
+    }
+    case CSR_HVIP: {
+      reg_t mask = MIP_VS_MASK;
+      state.mip = (state.mip & ~mask) | (val & mask);
+      break;
+    }
+    case CSR_HTINST:
+      state.htinst = val;
+      break;
+    case CSR_HGATP: {
+      reg_t reg_val = 0;
+      reg_t rv64_ppn_mask = (reg_t(1) << (MAX_PADDR_BITS - PGSHIFT)) - 1;
+      mmu->flush_tlb();
+      if (max_xlen == 32)
+        reg_val = val & (HGATP32_PPN | HGATP32_MODE);
+      if (max_xlen == 64 && (get_field(val, HGATP64_MODE) == HGATP_MODE_OFF ||
+                             get_field(val, HGATP64_MODE) == HGATP_MODE_SV39X4 ||
+                             get_field(val, HGATP64_MODE) == HGATP_MODE_SV48X4))
+        reg_val = val & (HGATP64_PPN | HGATP64_MODE | rv64_ppn_mask);
+      state.hgatp = reg_val;
+      break;
+    }
+    case CSR_VSSTATUS: {
+      reg_t mask = SSTATUS_VS_MASK;
+      mask |= (supports_extension('F') ? SSTATUS_FS : 0);
+      mask |= (supports_extension('V') ? SSTATUS_VS : 0);
+      mask |= (xlen == 64 ? SSTATUS64_SD : SSTATUS32_SD);
+      state.vsstatus = (state.vsstatus & ~mask) | (val & mask);
+      break;
+    }
+    case CSR_VSIE: {
+      reg_t mask = MIP_VS_MASK;
+      state.mie = (state.mie & ~mask) | ((val << 1) & mask);
+      break;
```

Comment:
hideleg should be considered here. I've confirmed this behavior with J. Hauser:
```
Second, as you say, hideleg should always be active.  (Although older
versions of the spec did accidentally say otherwise.)
```
Ditto for VSIP.

Reply:
I had considered HIDELEG for SIP and SIE access from VS-mode but forgot to consider it for VSIP and VSIE.

Thanks for catching. I will update.